rfcs/043-recording-deletions/README.md

```diff
@@ -0,0 +1,90 @@
+
+At time of writing, we have 1.4M deleted Works (44% of the 3.2M works in the index).
+Every time we reindex, we have to reprocess all of these records – almost certainly a non-trivial cost of a reindex.
+
+We should remove deleted source records from reindexes.
+
+## The proposal
+
+Currently we keep track of source records in a VHS.
+Every source record that has ever existed has an object in S3 and a pointer to that object in DynamoDB:
+
+![Two blue DynamoDB tables (left) pointing to a red S3 bucket (right)](table_structure.png)
+
+When we reindex, we read every item from the DynamoDB table and send that to the pipeline.
+
+The proposed approach: we have two DynamoDB tables with an identical structure:
+
+-   A "live" table where we write new records in the adapter, and which is used by the catalogue pipeline and the reindexer.
+    This is a continuation of the deleted table.
+
+-   A "deleted" table.
+    We move pointers to this table for source records that we don't want in the pipeline.
+    This table is ignored by the catalogue pipeline and the reindexer.
+
+We'd keep the same set of records overall, but now they're split over two tables.
+Over time, we move records between the two tables as we want to expose or hide them in the catalogue pipeline.
+
+We'd have per-source scripts with two options:
+
+-   Move all records deleted before a given date from "live" to "deleted"
+-   Move specific records from "deleted" to "live"
+
+This is intentionally a lossless, reversible process.
+We're not permanently deleting any data, and we could revert it later if there are unforeseen issues.
+
+## Benefits
+
+-   **We'd save money on reindexes.**
+    Deleted works account for 44% of the index, which means they're probably a substantial reindex cost.
+
+    (I don't have exact numbers, but a reindex costs ~$150.
+    Deleted Works are smaller than visible Works so they're a bit cheaper, but a lot of the costs are the same.
+    I'd guess we're spending ~$50 per reindex to process deleted works.)
+
+-   **Reindexes would go faster.**
+    Having less records means less work to do in a reindex, so they should finish faster.
+
+## Risks and mitigations
+
+-   **410 Gone would become 404 Not Found for any Works that we moved.**
+    If we don't put deleted Works in the final Elasticsearch index, the API will serve them as 404 Not Found instead of 410 Gone.
+    This is potentially annoying for anybody who has a URL to a work that was previously available.
+
+    As a first pass, we can avoid this by only removing Works that were deleted before the new catalogue API/website -- and so never had public URLs with the current URL scheme.
+    If you filter to anything deleted before 1 Jan 2017, there are 876k Works (27% of the 3.2M total Works), and so still a significant saving.
+
+    After that, we could move records to the "deleted" table on a rolling window -- say, anything deleted more than a year previously.
```

Comment:
This is the piece that still concerns me in the sense of we still land up with 404 that should be 410s if they're a year old. Maybe I am being a pedant about this, but it feels we could have an publically accessible version of 410s somehow and this statement seems to suggest we just shouldn't.

Reply:
It could even mean that we have a separate pipeline or similar setup, which could even put deleted works into a separate index, which we can search.